### PR TITLE
chore: bump network controller 21.0.0

### LINKED ISSUE #11229: [Network-controller-upgrade ] Update Controller Version to v21
This task involves updating the Network Controller to version 21 in MetaMask Mobile. The primary objective is to migrate the state structure from the current version to the new version, ensuring compatibility with the latest features and improvements.
**Current State (Version 20):**
```
{
  "selectedNetworkClientId": "673a4523-3c49-47cd-8d48-68dfc8a47a9c",
  "networksMetadata": {
    "mainnet": {
      "status": "available",
      "EIPS": {
        "1559": true
      }
    },
    "linea-mainnet": {
      "status": "available",
      "EIPS": {
        "1559": true
      }
    },
    "d740f3bf-4b3a-4ce5-80f0-df13071ae5fc": {
      "status": "available",
      "EIPS": {
        "1559": true
      }
    },
    "673a4523-3c49-47cd-8d48-68dfc8a47a9c": {
      "status": "available",
      "EIPS": {
        "1559": true
      }
    }
  },
  "networkConfigurations": {
    "673a4523-3c49-47cd-8d48-68dfc8a47a9c": {
      "id": "673a4523-3c49-47cd-8d48-68dfc8a47a9c",
      "rpcUrl": "https://bsc-dataseed1.binance.org/",
      "chainId": "0x38",
      "ticker": "BNB",
      "nickname": "BNB Chain",
      "rpcPrefs": {
        "blockExplorerUrl": "https://bscscan.com"
      }
    },
    "any-other-id": {
      "id": "any-other-id",
      "rpcUrl": "https://bsc-mainnet.public.blastapi.io",
      "chainId": "0x38",
      "ticker": "BNB",
      "nickname": "BNB Chain",
      "rpcPrefs": {
        "blockExplorerUrl": "https://bscscan.com"
      }
    }
  }
}
```

**Target State (Version 21):**
```
{
  "networkConfigurationsByChainId": {
    "0x38": {
      "chainId": "0x38",
      "rpcEndpoints": [
        {
          "networkClientId": "673a4523-3c49-47cd-8d48-68dfc8a47a9c",
          "url": "https://bsc-dataseed1.binance.org/",
          "type": "custom",
          "name": "BNB Chain"
        },
        {
          "networkClientId": "any-other-id",
          "url": "https://bsc-mainnet.public.blastapi.io",
          "type": "custom",
          "name": "BNB Chain"
        }
      ],
      "defaultRpcEndpointIndex": 0,
      "blockExplorerUrls": ["https://bscscan.com"],
      "defaultBlockExplorerUrlIndex": 0,
      "name": "Binance Smart Chain",
      "nativeCurrency": "BNB"
    }
  },
  "networksMetadata": {
    "mainnet": {
      "status": "available",
      "EIPS": { "1559": true }
    },
    "linea-mainnet": {
      "status": "available",
      "EIPS": { "1559": true }
    },
    "d740f3bf-4b3a-4ce5-80f0-df13071ae5fc": {
      "status": "available",
      "EIPS": { "1559": true }
    },
    "673a4523-3c49-47cd-8d48-68dfc8a47a9c": {
      "status": "available",
      "EIPS": { "1559": true }
    }
  },
  "selectedNetworkClientId": "673a4523-3c49-47cd-8d48-68dfc8a47a9c"
}
```

**Key Changes:**

**_Network Configurations:_**
The `networkConfigurations` object is replaced with `networkConfigurationsByChainId`.
Each chain ID now maps to an object containing `rpcEndpoints`, `blockExplorerUrls`, and other metadata.
`rpcEndpoints` is an array of RPC endpoint objects, each containing networkClientId, url, type, and name.
`defaultRpcEndpointIndex` and `defaultBlockExplorerUrlIndex` are added to specify the default RPC endpoint and block explorer URL.

**_Selected Network Client ID:_**
The selectedNetworkClientId remains the same but is now part of the updated state structure.
Networks Metadata:

**The networksMetadata** structure remains unchanged, ensuring compatibility with existing metadata.

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**
This PR updates the network-controller from version 20.0.0 to 21.0.0. The migration introduces changes in the handling of network configurations, particularly regarding how network data is structured and referenced.

Key Changes:
Consolidation of Network Configurations:

Network configurations in v21 are now organized by chainId in a new networkConfigurationsByChainId structure, replacing the networkConfigurations section in v20.
The new structure supports multiple RPC endpoints per network, as seen in the list of rpcEndpoints, with support for Infura-based connections (e.g., https://mainnet.infura.io/v3/{infuraProjectId}).

## **Remark**

> As part of the update to version 21 (v21) of the network controller, we are making a key change in how networks are handled. Previously, networks could have the same chain ID but different RPC URLs. With v21, each network will now have a unique chain ID that can include multiple RPC URLs.

> To support this, the UI will also be updated—for example, the RPC URL field in the network form will become a dropdown menu instead of a simple text input.

> These changes have been implemented behind a feature flag, which will remain in place until the v21 upgrade is finalized.

> To complete this migration, we have 3 PRs that need to be reviewed and merged together:

> The first PR upgrades the network controller to v21.
> The second PR adds the migration script and the necessary UI for migration.
> The third PR removes the feature flag, making the v21 changes active.
> For the review process, it's recommended to start with the third PR (removing the feature flag), then proceed to the > second and first PRs. This order ensures a smooth transition and proper implementation of the changes.

## **Related PRS**

- https://github.com/MetaMask/metamask-mobile/pull/11622 ( contain the migration script + modal )
- https://github.com/MetaMask/metamask-mobile/pull/11436 ( contain the remove of the feature flag , the new UI is visible )

## **Related issues**

Fixes: #11229 

## **Manual testing steps**

- Verify that networkConfigurations has been replaced by networkConfigurationsByChainId.
- Check that each network in networkConfigurationsByChainId is identified by its chainId.
- Ensure that all rpcEndpoints are correctly listed for each network in networkConfigurationsByChainId.
- Validate that the defaultRpcEndpointIndex points to a valid and functional RPC endpoint for each network.
- Check that Infura-based RPC URLs (e.g., Mainnet, Goerli, Sepolia, Linea networks) are properly formatted with the correct {infuraProjectId}.
- Verify that networks like Linea Goerli, Linea Sepolia, and Linea Mainnet are listed under networkConfigurationsByChainId with their correct chainId values.
- Ensure that any unused or outdated custom network configurations (e.g., BNB Chain from v20) have been removed.
- Check that all network names (e.g., "Mainnet", "Goerli", "Sepolia") in networkConfigurationsByChainId match their respective chainId values.
- Validate that nativeCurrency fields are correctly set for each network in networkConfigurationsByChainId (e.g., ETH for Mainnet, SepoliaETH for Sepolia).
- Ensure that each network's blockExplorerUrls is present and correctly configured where applicable.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by legobeat** [comment]
Related:
- #9930
- #9274
- #9925

File: app/components/Views/NFTAutoDetectionModal/NFTAutoDetectionModal.test.tsx [hunk 1]
```diff
@@ -33,6 +35,15 @@ const initialState = {
...
         displayNftMedia: true,
       },
+      NetworkController: {
+        ...mockNetworkState({
+          chainId: CHAIN_IDS.MAINNET,
+          id: 'mainnet',
+          nickname: 'Ethereum Mainnet',
+          ticker: 'ETH',
+          blockExplorerUrl: 'https://goerli.lineascan.build',
+        }),
```
[the remark this answers is not in the document]
Comment (salimtb) [reply app/components/Views/NFTAutoDetectionModal/NFTAutoDetectionModal.test.tsx:45]: awesome idea thnx @darkwing , just added to my todo list

File: app/components/UI/Ramp/Views/NetworkSwitcher/NetworkSwitcher.tsx [hunk 2]
```diff
@@ -159,11 +159,16 @@ function NetworkSwitcher() {
...
+        const [, { nativeCurrency: ticker }] = entry;
+
+        const { networkClientId } =
+          networkConfiguration?.rpcEndpoints?.[
+            networkConfiguration.defaultRpcEndpointIndex
+          ] ?? {};
 
         CurrencyRateController.updateExchangeRate(ticker);
-        NetworkController.setActiveNetwork(networkConfigurationId);
+        // TODO: fix it here setActiveNetwork
```
Comment (salimtb) [inline app/components/UI/Ramp/Views/NetworkSwitcher/NetworkSwitcher.tsx:170]: Todo: remove this comment

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 3]
```diff
@@ -128,7 +129,7 @@ const NetworkSelector = () => {
   const sheetRef = useRef<BottomSheetRef>(null);
   const showTestNetworks = useSelector(selectShowTestNetworks);
 
-  const providerConfig: ProviderConfig = useSelector(selectProviderConfig);
+  const providerConfig = useSelector(selectProviderConfig);
```
Comment (darkwing) [inline app/components/Views/NetworkSelector/NetworkSelector.tsx:132]: Does this type need to be removed?

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 4]
```diff
@@ -232,27 +241,45 @@ const NetworkSelector = () => {
...
+        },
+      ] = entry;
+
+      const networkConfigurationId =
+        NetworkController.findNetworkClientIdByChainId(
+          chainId as `0x${string}`,
+        );
 
       if (domainIsConnectedDapp && process.env.MULTICHAIN_V1) {
+        // todo: check with the author
```
Comment (darkwing) [inline app/components/Views/NetworkSelector/NetworkSelector.tsx:266]: Noting `todo`

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 5]
```diff
@@ -232,27 +241,45 @@ const NetworkSelector = () => {
...
         SelectedNetworkController.setNetworkClientIdForDomain(
           origin,
           networkConfigurationId,
         );
       } else {
         CurrencyRateController.updateExchangeRate(ticker);
-        NetworkController.setActiveNetwork(networkConfigurationId);
+
+        const { networkClientId } =
+          rpcEndpoints?.[defaultRpcEndpointIndex] ?? {};
```
Comment (darkwing) [inline app/components/Views/NetworkSelector/NetworkSelector.tsx:275]: Does not having this `networkClientId` get us in trouble?  If not, should there be an `if(networkClientId )` before `await NetworkController.setActiveNetwork(networkClientId);`?

File: app/lib/ppom/ppom-util.test.ts [hunk 6]
```diff
@@ -31,24 +35,6 @@ jest.mock('../../core/Engine', () => ({
     PPOMController: {
       usePPOM: jest.fn(),
     },
-    // NetworkController: {
```
Comment (darkwing) [inline app/lib/ppom/ppom-util.test.ts:34]: Commented code needs removal.

File: app/lib/ppom/ppom-util.ts [hunk 7]
```diff
@@ -69,6 +69,7 @@ async function validateRequest(req: PPOMRequest, transactionId?: string) {
   } = NetworkController.getNetworkClientById(
     NetworkController.state?.selectedNetworkClientId,
   );
+
```
Comment (darkwing) [inline app/lib/ppom/ppom-util.ts:72]: Empty file change makes the PR one file size larger.  We can remove this.

File: app/selectors/networkController.ts [hunk 8]
```diff
@@ -26,55 +84,51 @@ export const selectSelectedNetworkClientId = createSelector(
...
+
+    // Return default provider config (mainnet) if no matching network is found
+    return getDefaultProviderConfig();
   },
 );
 
 export const selectTicker = createSelector(
   selectProviderConfig,
-  (providerConfig: ProviderConfig) => providerConfig?.ticker,
+  (providerConfig) => providerConfig?.ticker,
```
Comment (darkwing) [inline app/selectors/networkController.ts:114]: Is removing all of these type checks problematic?  Feels like it could be so.

File: app/store/migrations/023.test.ts [hunk 9]
```diff
@@ -60,6 +60,9 @@ describe('Migration #23', () => {
         engine: {
           backgroundState: {
             AddressBookController: null,
+            NetworkController: {
+              networkConfigurations: {},
```
Comment (darkwing) [inline app/store/migrations/023.test.ts:64]: Is touching old migrations problematic?

File: app/store/migrations/029.ts [hunk 10]
```diff
@@ -63,8 +62,9 @@ export default async function migrate(stateAsync: unknown) {
   }
 
   const networkControllerState = state.engine.backgroundState.NetworkController;
-  const newNetworkControllerState = state.engine.backgroundState
-    .NetworkController as NetworkState;
+  const newNetworkControllerState =
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    state.engine.backgroundState.NetworkController as any;
```
Comment (darkwing) [inline app/store/migrations/029.ts:67]: Can we create a type for this?

File: app/component-library/components-temp/CellSelectWithMenu/CellSelectWithMenu.tsx [hunk 11]
```diff
@@ -46,12 +47,15 @@ const CellSelectWithMenu = ({
...
-          {...avatarProps}
-        />
+        {withAvatar ? (
+          <Avatar
+            style={styles.avatar}
+            testID={CellModalSelectorsIDs.BASE_AVATAR}
+            size={DEFAULT_CELLBASE_AVATAR_SIZE}
+            {...avatarProps}
+          />
+        ) : null}
```
Comment (darkwing) [inline app/component-library/components-temp/CellSelectWithMenu/CellSelectWithMenu.tsx:57]: Is this required for this PR?  Should it be a separate change?

File: app/components/UI/NetworkModal/index.tsx [hunk 12]
```diff
@@ -183,56 +192,152 @@ const NetworkModals = (props: NetworkProps) => {
...
     const url = new URLPARSE(rpcUrl);
+    const existingNetwork = networkConfigurationByChainId[chainId];
+
     CurrencyRateController.updateExchangeRate(ticker);
-    !isprivateConnection(url.hostname) && url.set('protocol', 'https:');
-    NetworkController.upsertNetworkConfiguration(
-      {
-        rpcUrl: url.href,
+
+    if (!isprivateConnection(url.hostname)) {
```
Comment (darkwing) [inline app/components/UI/NetworkModal/index.tsx:318]: I think this should be renamed `isPrivateConnection`

File: app/components/Views/Notifications/Details/Footers/BlockExplorerFooter.tsx [hunk 13]
```diff
@@ -32,7 +32,8 @@ export default function BlockExplorerFooter(props: BlockExplorerFooterProps) {
     const hexChainId = toHex(props.chainId);
     return Object.values(networkConfigurations).find(
       (networkConfig) => networkConfig.chainId === hexChainId,
-    )?.rpcPrefs?.blockExplorerUrl;
+      // todo : fix the multi block explorer
+    )?.blockExplorerUrls?.[0];
```
Comment (darkwing) [inline app/components/Views/Notifications/Details/Footers/BlockExplorerFooter.tsx:36]: Fix?

File: app/components/hooks/useTokenBalancesController/useTokenBalancesController.test.tsx [hunk 14]
```diff
@@ -76,8 +76,6 @@ describe('useTokenBalancesController()', () => {
     // starting from fresh store with the initial state
     testStore = createStore(testBalancesReducer, cloneDeep(mockInitialState));
 
-    // console.log('testStore', JSON.stringify(testStore.getState()));
```
Comment (darkwing) [inline app/components/hooks/useTokenBalancesController/useTokenBalancesController.test.tsx:79]: Removing this change will make the PR smaller.

File: app/components/Views/NetworkSelector/NetworkSelector.tsx (hunk 4 above)
Comment (salimtb) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:266]: already verified with author

File: app/components/Views/NetworkSelector/NetworkSelector.tsx (hunk 5 above)
Comment (salimtb) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:275]: Thank you for the feedback!

We have the same logic implemented in the extension, and it works because networkClientId cannot be null or undefined. Additionally, the setActiveNetwork method already handles this case internally, so there is no risk of issues arising from a missing or undefined networkClientId.

Please let me know if you need further clarification!

File: app/store/migrations/023.test.ts (hunk 9 above)
Comment (salimtb) [reply app/store/migrations/023.test.ts:64]: No, this isn't an issue because the migration file is still designed to handle the old state and behaves as expected. The change was added solely for testing purposes.

File: app/store/migrations/029.ts (hunk 10 above)
Comment (salimtb) [reply app/store/migrations/029.ts:67]: done

File: app/store/migrations/023.test.ts (hunk 9 above)
Comment (salimtb) [reply app/store/migrations/023.test.ts:64]: the migration initially worked with the state located at `../util/test/initial-root-state` , which referred to the old state that included the `networkConfigurations` object. However, with the upgrade to v21, this property no longer exists. Therefore, our migration script needs to take the old state as input, which is the reason for this change.

File: app/selectors/networkController.ts (hunk 8 above)
Comment (salimtb) [reply app/selectors/networkController.ts:114]: we are sending a string and not a `ProviderConfig` object , this can be replaced by `string` instead

File: yarn.lock [hunk 15]
```diff
@@ -5055,7 +5083,7 @@
     immer "^9.0.6"
     uuid "^8.3.2"
 
-"@metamask/network-controller@^20.0.0", "@metamask/network-controller@^20.1.0":
+"@metamask/network-controller@^20.0.0":
```
Comment (legobeat) [inline yarn.lock:5086]: `@metamask/network-controller@20.1.0` is still used by:
- `@metamask/gas-fee-controller@^19` via `@metamask/transaction-controller@^35`
- `@metamask/polling-controller@^9` via `@metamask/gas-fee-controller@^19`
- `@metamask/smart-transactions-controller@^13`
- `@metamask/transaction-controller@^35`
In addition to existing references to `@metamask/network-controller@^19`, `@metamask/network-controller@^18`, & `@metamask/network-controller@^17`.
Is this intended and fine with regards to controller compatibility?

Comment (legobeat) [reply yarn.lock:5086]: Related peer-dependency vialoations from `yarn install`:
```
 warning " > @metamask/assets-controllers@30.0.0" has incorrect peer dependency "@metamask/network-controller@^18.1.2".
 warning " > @metamask/gas-fee-controller@18.0.0" has incorrect peer dependency "@metamask/network-controller@^19.0.0".
warning " > @metamask/selected-network-controller@15.0.2" has incorrect peer dependency "@metamask/network-controller@^19.0.0".
warning "@metamask/smart-transactions-controller > @metamask/transaction-controller@34.0.0" has incorrect peer dependency "@metamask/network-controller@^19.0.0".
warning " > @metamask/swaps-controller@9.0.12" has incorrect peer dependency "@metamask/network-controller@^3.0.0".
warning "@metamask/swaps-controller > @metamask/gas-fee-controller@12.0.0" has incorrect peer dependency "@metamask/network-controller@^17.1.0".
warning " > @metamask/transaction-controller@35.0.0" has incorrect peer dependency "@metamask/network-controller@^20.0.0".
```

Comment (salimtb) [reply yarn.lock:5086]: hey @legobeat 
Yes, it is fine, and the compatibility across these versions is correct. The `@metamask/network-controller` versions referenced by the various dependencies do not pose any issues in terms of compatibility. 

The only exception is the `@metamask/transaction-controller`, where we are now using a patch to ensure compatibility between versions and fix any issues that arose with the network-controller integration. This patch resolves specific problems and maintains expected behavior within the transaction-controller.

In summary, while multiple versions are referenced, they do not introduce compatibility concerns, and the necessary patch for the transaction-controller ensures everything works as expected.

File: app/core/Engine.ts [hunk 16]
```diff
@@ -1333,6 +1333,7 @@ class Engine {
       // but only breaking change is Node version and bumped dependencies
       getNetworkClientRegistry:
         networkController.getNetworkClientRegistry.bind(networkController),
+      // @ts-expect-error TODO: resolve types
```
Comment (MajorLift) [inline app/core/Engine.ts:1336]: Following up on https://github.com/MetaMask/metamask-mobile/pull/11292#discussion_r1777341712, this is exactly the kind of scenario that `peerDependencies` are meant to help us avoid.

Looking at the error message here:
> Property 'networkConfigurations' is missing in type 'import(".../salim/bump-network-controller-21.0.0/node_modules/@metamask/network-controller/dist/NetworkController").NetworkState' but required in type 'import(".../salim/bump-network-controller-21.0.0/node_modules/@metamask/transaction-controller/node_modules/@metamask/network-controller/dist/types/NetworkController").NetworkState'.ts(2741)

There's no way to tell whether the missing `networkConfigurations` property will break `TransactionController` at runtime without looking into implementation details.

File: yarn.lock (hunk 15 above)
Comment (salimtb) [reply yarn.lock:5086]: hey @MajorLift ,
ty for your comment, the only solution to align with the strict peerDependencies requirements mentioned would be to upgrade the transaction controller. However, this is outside the scope of this PR, which is already large and would become even more complex if we included the transaction controller upgrade. If upgrading the transaction controller is necessary, it could potentially block progress on the network controller upgrade.

That being said, is there any other solution we can explore that avoids the need for upgrading the transaction controller at this stage?

File: app/core/Engine.ts (hunk 16 above)
Comment (salimtb) [reply app/core/Engine.ts:1336]: Thank you for the follow-up. You're correct that this is precisely the kind of issue that peerDependencies are intended to prevent.

Regarding the error message you pointed out:
```
Property 'networkConfigurations' is missing in type 'import(".../salim/bump-network-controller-21.0.0/node_modules/@metamask/network-controller/dist/NetworkController").NetworkState' but required in type 'import(".../salim/bump-network-controller-21.0.0/node_modules/@metamask/transaction-controller/node_modules/@metamask/network-controller/dist/types/NetworkController").NetworkState'.ts(2741)
```

This suggests a potential incompatibility between the two controller versions. The missing networkConfigurations property could indeed cause issues for TransactionController at runtime, but it's difficult to confirm without delving deeper into the implementation details.

As previously mentioned, upgrading the TransactionController would be the most straightforward solution, but that is out of scope for this PR. If this upgrade is required, it could act as a blocker for moving forward with the network controller update.

Do you see any alternative solutions that could address this issue while avoiding the need for an immediate upgrade of the TransactionController?

Comment (cryptodev-2s) [reply app/core/Engine.ts:1336]: @MajorLift currently looking on the implementation and will upgrade it if needed

File: yarn.lock (hunk 15 above)
Comment (salimtb) [reply yarn.lock:5086]: Transaction controller has been upgraded to v37, which resolve the peer dependency issue, i'll mark this comment as resolved

Comment (MajorLift) [reply yarn.lock:5086]: Nice! Thanks for being receptive to making this change. 

Your concern about PR scope is definitely valid. This could be mitigated by extracting each controller upgrade into separate PRs. The downstream controller(s) could then target the upstream controller as their base branch, or both could be merged into a third staging branch. This would keep each PR small enough to be reviewable while also satisfying the peer dependency requirements.

This isn't a request for this PR though, just a thought for future reference.

**Comment by MajorLift** [comment]
Looks like there are a few unused imports that should be removed:
https://sonarcloud.io/project/issues?pullRequest=11292&id=metamask-mobile&open=AZIpDbcZclZm4itKzMCp

**Comment by legobeat** [comment]
@salimtb Needs conflict-resolution with `main`

File: app/components/UI/NetworkModal/index.tsx [hunk 17]
```diff
@@ -7,7 +7,7 @@ import NetworkDetails from './NetworkDetails';
 import NetworkAdded from './NetworkAdded';
 import Engine from '../../../core/Engine';
 import {
-  isprivateConnection,
+  isPrivateConnection,
```
[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply app/components/UI/NetworkModal/index.tsx:10]: yes extracted to https://github.com/MetaMask/metamask-mobile/pull/11656

File: app/store/migrations/051.test.ts [hunk 18]
```diff
@@ -111,6 +111,7 @@ describe('Migration #51', () => {
         engine: {
           backgroundState: {
             NetworkController: {
+              networkConfigurations: {},
```
[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply app/store/migrations/051.test.ts:114]: yes otherwise this test was failing

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 19]
```diff
@@ -895,10 +940,12 @@ const NetworkSelector = () => {
               )}
               iconName={IconName.Edit}
               onPress={() => {
-                navigate(Routes.ADD_NETWORK, {
-                  shouldNetworkSwitchPopToWallet: false,
-                  shouldShowPopularNetworks: false,
-                  network: showNetworkMenuModal.networkTypeOrRpcUrl,
+                sheetRef.current?.onCloseBottomSheet(() => {
```
[the remark this answers is not in the document]
Comment (salimtb) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:944]: The change was made to ensure a smoother user experience. By delaying the navigation until after the bottom sheet has fully closed, we avoid potential issues with the UI transitioning abruptly while the bottom sheet is still in motion. This ensures that the user’s focus remains on one action at a time, preventing any overlap between the bottom sheet closing animation and the navigation process. Essentially, it makes the transition cleaner and more fluid.

File: app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js
[the remark this answers is not in the document]
Comment (salimtb) [reply app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:1]: yes this is possible , i'll move them to the third PR

File: app/util/networks/index.js [hunk 20]
```diff
@@ -405,6 +403,10 @@ export const getNetworkNameFromProviderConfig = (providerConfig) => {
   let name = strings('network_information.unknown_network');
   if (providerConfig.nickname) {
     name = providerConfig.nickname;
+  } else if (providerConfig.chainId === NETWORKS_CHAIN_ID.MAINNET) {
```
[the remark this answers is not in the document]
Comment (salimtb) [reply app/util/networks/index.js:406]: this condition is responsible to display the network name on the select network nav bar , in mobile the network name for mainnet must always be `Ethereum Main Network` , I didn't want to change this in order to continue having the same behavior.
Screenshot 2024-10-09 at 11 11 38

File: app/store/migrations/051.test.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (salimtb) [reply app/store/migrations/051.test.ts:114]: This change is needed because the updated state in `util/test/initial-root-state` no longer includes the `networkConfigurations` property ( because the network controller has been moved to v21 ). As a result, the migration script fails since it relies on the old state format that contains this property.

File: app/components/UI/AccountInfoCard/index.test.tsx [hunk 21]
```diff
@@ -48,20 +50,13 @@ const mockInitialState: DeepPartial<RootState> = {
...
-            rpcUrl: 'http://localhost/v3/',
-            chainId: '0xaa36a7',
-            ticker: 'ETH',
-            nickname: 'sepolia',
-            rpcPrefs: {
-              blockExplorerUrl: 'https://etherscan.com',
-            },
-          },
-        },
+        ...mockNetworkState({
```
[the remark this answers is not in the document]
Comment (salimtb) [reply app/components/UI/AccountInfoCard/index.test.tsx:53]: I'm not sure how that would reduce the size of this PR. If we don't use this function, we'll have to go through the networkController state object and update it ( the one i just delete and replaced with this function ) instead. Unless I've misunderstood your point here

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 22]
```diff
@@ -382,8 +404,8 @@ const NetworkSelector = () => {
             imageSource: images.ETHEREUM,
             size: AvatarSize.Sm,
           }}
-          isSelected={chainId === selectedChainId && !providerConfig.rpcUrl}
-          onPress={() => onNetworkChange(MAINNET)}
+          isSelected={chainId === selectedChainId && !providerConfig?.rpcUrl}
```
[the remark this answers is not in the document]
Comment (legobeat) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:439]: Separate q: What's the significance/difference of using `providerConfig.rpcUrl` instead of `rpcUrl` here? I think a one-line comment on that would help make this a lot more clear.

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 23]
```diff
@@ -470,85 +520,96 @@ const NetworkSelector = () => {
...
+        defaultRpcEndpointIndex,
+      } = networkConfiguration;
+      if (
+        !chainId ||
+        isTestNet(chainId) ||
+        isMainNet(chainId) ||
+        chainId === CHAIN_IDS.LINEA_MAINNET ||
+        chainId === CHAIN_IDS.GOERLI
+      ) {
+        return null;
```
Comment (legobeat) [inline app/components/Views/NetworkSelector/NetworkSelector.tsx:544]: Why is `renderRpcNetworks` skipped except for this set of hardcoded known chains? Is there any difference in functionality between the two cases?

Comment (salimtb) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:544]: Here's a clearer explanation of the same concept:

We skip rendering certain networks in the `if` block because we need to handle them separately to manage their order and display settings. Specifically, test networks (testnets) need to appear at the bottom of the list, and their visibility should be controlled by a toggle option (whether to show or hide them).

For the primary networks, *Mainnet* and *Linea*, they should always appear as the first and second items in the network list by default. 

- *Mainnet* is rendered using the `renderMainnet` function.
- *Linea* is rendered with the `renderLineaMainnet` function.
- All other networks are rendered through the `renderRpcNetworks` function.
- Test networks (testnets) are rendered using the `renderOtherNetworks` function.

To manage the display order and ensure proper handling, we exclude these specific networks from the `networkConfiguration` state, allowing us to render them with greater control in the desired order.

**Comment by salimtb** [comment]
This pull request was merged into [this feature branch](https://github.com/MetaMask/metamask-mobile/pull/11705), so I need approval on that feature branch. I'm keeping the PR open to make the review process easier.
